src/modules/hyprland/workspaces.cpp

```diff
@@ -0,0 +1,167 @@
+#include "modules/hyprland/workspaces.hpp"
+
+#include <json/value.h>
+#include <spdlog/spdlog.h>
+
+#include <algorithm>
+#include <charconv>
+#include <string>
+
+namespace waybar::modules::hyprland {
+
+Workspaces::Workspaces(const std::string &id, const Bar &bar, const Json::Value &config)
+    : AModule(config, "workspaces", id, false, false),
+      bar_(bar),
+      box_(bar.vertical ? Gtk::ORIENTATION_VERTICAL : Gtk::ORIENTATION_HORIZONTAL, 0) {
+  Json::Value config_format = config["format"];
+
+  format_ = config_format.isString() ? config_format.asString() : "{id}";
+  with_icon_ = format_.find("{icon}") != std::string::npos;
+
+  if (with_icon_ && icons_map_.empty()) {
+    Json::Value format_icons = config["format-icons"];
+    for (std::string &name : format_icons.getMemberNames()) {
+      icons_map_.emplace(name, format_icons[name].asString());
+    }
+
+    icons_map_.emplace("", "");
+  }
+
+  box_.set_name("workspaces");
+  if (!id.empty()) {
+    box_.get_style_context()->add_class(id);
+  }
+  event_box_.add(box_);
+  modulesReady = true;
+  if (!gIPC.get()) {
+    gIPC = std::make_unique<IPC>();
+  }
+
+  init();
+
+  gIPC->registerForIPC("workspace", this);
+  gIPC->registerForIPC("createworkspace", this);
+  gIPC->registerForIPC("destroyworkspace", this);
+}
+
+auto Workspaces::update() -> void {
+  std::lock_guard<std::mutex> lock(mutex_);
+  for (Workspace &workspace : workspaces_) {
+    workspace.set_active(workspace.id() == active_workspace_id);
+
+    std::string &workspace_icon = icons_map_[""];
+    if (with_icon_) {
+      workspace_icon = workspace.select_icon(icons_map_);
+    }
+
+    workspace.update(format_, workspace_icon);
+  }
+
+  AModule::update();
+}
+
+void Workspaces::onEvent(const std::string &ev) {
+  std::string eventName(begin(ev), begin(ev) + ev.find_first_of('>'));
+  std::string payload = ev.substr(eventName.size() + 2);
+  if (eventName == "workspace") {
+    std::from_chars(payload.data(), payload.data() + payload.size(), active_workspace_id);
+  } else if (eventName == "destroyworkspace") {
+    int deleted_workspace_id;
+    std::from_chars(payload.data(), payload.data() + payload.size(), deleted_workspace_id);
+    workspaces_.erase(std::remove_if(workspaces_.begin(), workspaces_.end(),
+                                     [&](Workspace &x) { return x.id() == deleted_workspace_id; }));
+  } else if (eventName == "createworkspace") {
+    int new_workspace_id;
+    std::from_chars(payload.data(), payload.data() + payload.size(), new_workspace_id);
+    workspaces_.push_back(new_workspace_id);
+    Gtk::Button &new_workspace_button = workspaces_.back().button();
+    box_.pack_end(new_workspace_button, false, false);
+    sort_workspaces();
+    new_workspace_button.show_all();
```

Comment:
The segfault happens here, when showing the new button.

I added some logging to that function, I put the `Workspaces: showing new workspace` line right before `show_all();`, this is what happens:

```
[2023-07-01 11:50:12.086] [info] Bar configured (width: 2560, height: 32) for output: eDP-1
[2023-07-01 11:50:13.459] [info] Workspaces: createworkspace>>12
[2023-07-01 11:50:13.459] [info] Workspaces: new workspace 12
[2023-07-01 11:50:13.460] [info] Workspaces: showing new workspace 12
[2023-07-01 11:50:13.463] [info] Workspaces: workspace>>12
[2023-07-01 11:50:13.523] [info] Workspaces: createworkspace>>13
[2023-07-01 11:50:13.524] [info] Workspaces: new workspace 13
[2023-07-01 11:50:13.524] [info] Workspaces: showing new workspace 13
[2023-07-01 11:50:13.527] [info] Workspaces: workspace>>13
[2023-07-01 11:50:13.528] [info] Workspaces: destroyworkspace>>12
[2023-07-01 11:50:14.253] [info] Workspaces: workspace>>11
[2023-07-01 11:50:14.254] [info] Workspaces: destroyworkspace>>13
[2023-07-01 11:50:14.767] [info] Workspaces: createworkspace>>12
[2023-07-01 11:50:14.767] [info] Workspaces: new workspace 12
[2023-07-01 11:50:14.767] [info] Workspaces: showing new workspace 12
[1]    437035 segmentation fault  ./build-test/waybar --config config.jsonc
```
Only workspace 11 has windows on it, the others are empty.

Reply:
It is not about 11+ workspace. I was able to reproduce it by just switching from 4 to 5 fast.
This issue happens because of saving refs from vector I suppose.

Reply:
I was able to reproduce with just 2 hits with next log:
```
[2023-07-01 14:06:25.634] [warning] Requested height: 21 is less than the minimum height: 25 required by the modules
[2023-07-01 14:06:25.634] [info] Bar configured (width: 3840, height: 25) for output: DP-2
[2023-07-01 14:06:25.635] [debug] hyprland IPC received openlayer>>waybar
[2023-07-01 14:06:28.444] [debug] hyprland IPC received createworkspace>>7
[2023-07-01 14:06:28.444] [debug] sort when 7 created
[2023-07-01 14:06:28.444] [debug] set workspace 1 on pos 0
[2023-07-01 14:06:28.444] [debug] set workspace 2 on pos 1
[2023-07-01 14:06:28.444] [debug] set workspace 3 on pos 2
[2023-07-01 14:06:28.444] [debug] set workspace 7 on pos 3
[2023-07-01 14:06:28.444] [debug] try to show all after 7 created
[2023-07-01 14:06:28.445] [debug] Task (0) ./waybar -l debug ~/D/P/W/build [Alacritty] <amif> changed
[2023-07-01 14:06:28.445] [debug] hyprland IPC received activewindow>>,
[2023-07-01 14:06:28.446] [debug] hyprland IPC received activewindowv2>>,
[2023-07-01 14:06:28.447] [debug] hyprland IPC received workspace>>7
[2023-07-01 14:06:29.286] [debug] hyprland IPC received createworkspace>>6
[2023-07-01 14:06:29.286] [debug] sort when 6 created
[2023-07-01 14:06:29.286] [debug] set workspace 1 on pos 0
[2023-07-01 14:06:29.286] [debug] set workspace 2 on pos 1
[2023-07-01 14:06:29.286] [debug] set workspace 3 on pos 2
[2023-07-01 14:06:29.286] [debug] set workspace 6 on pos 3
[2023-07-01 14:06:29.286] [debug] set workspace 7 on pos 4
[2023-07-01 14:06:29.286] [debug] try to show all after 6 created
fish: Job 1, './waybar -l debug' terminated by signal SIGSEGV (Address boundary error)
```
for code:
```c++
void Workspaces::onEvent(const std::string &ev) {
  std::string eventName(begin(ev), begin(ev) + ev.find_first_of('>'));
  std::string payload = ev.substr(eventName.size() + 2);
  std::lock_guard<std::mutex> lg(mutex_);
  if (eventName == "workspace") {
    std::from_chars(payload.data(), payload.data() + payload.size(), active_workspace_id);
  } else if (eventName == "destroyworkspace") {
    int deleted_workspace_id;
    std::from_chars(payload.data(), payload.data() + payload.size(), deleted_workspace_id);
    auto workspace = std::find_if(workspaces_.begin(), workspaces_.end(),
                                  [&](Workspace &x) { return x.id() == deleted_workspace_id; });
    spdlog::debug("remove button when {} deleted", deleted_workspace_id);
    box_.remove(workspace->button());
    workspaces_.erase(workspace);
    spdlog::debug("erase workspace when {} deleted", deleted_workspace_id);
  } else if (eventName == "createworkspace") {
    int new_workspace_id;
    std::from_chars(payload.data(), payload.data() + payload.size(), new_workspace_id);
    workspaces_.push_back(new_workspace_id);
    auto &new_workspace_button =
        std::find_if(workspaces_.begin(), workspaces_.end(), [&](Workspace &x) {
          return x.id() == new_workspace_id;
        })->button();
    box_.pack_end(new_workspace_button, false, false);
    spdlog::debug("sort when {} created", new_workspace_id);
    sort_workspaces();
    spdlog::debug("try to show all after {} created", new_workspace_id);
    auto &new_workspace_button_after_sort =
        std::find_if(workspaces_.begin(), workspaces_.end(), [&](Workspace &x) {
          return x.id() == new_workspace_id;
        })->button();
    new_workspace_button_after_sort.show_all();
  }

  dp.emit();
}
```

Reply:
It's strange that it also happens when not calling `sort_workspaces()`, so we're not changing the order in the vector of refs.

Reply:
I can assume there is a problem with that other thread is calling event handling. Maybe changing event handling to sway way will help.

Reply:
Adding a lock guard in the onEvent function does not fix the issue. If it was another thread calling the event handler, would this not fix it?

Reply:
Yes, another thread isn't a problem. Sway IPC works almost the same, but just more complicated way.
As for now, I have lack of ideas why it is happening.

Reply:
I figured it out!